### PR TITLE
Update mlc_chat_cli pipeline without kv cache

### PR DESCRIPTION
**Pull request by vvchernov**
It is not necessary that transformer-based models use kv cache (example, mpt-7b-instruct and possibly implemented rwkv). But mlc_chat_cli pipeline requires create_kv_cache function. There is fix allowing to not usekv_cache, it does not affect usual pipeline for llama

cc @tqchen

### CODE REVIEW

**Comment by tqchen** [comment]
I am a bit surprised that mpt does not require kv cache. Even if the original model do not use kv, we should build a version with kvcache for efficiency reasons

**Comment by yzh119** [comment]
@vvchernov I think mpt still relies on kv-cache (https://huggingface.co/mosaicml/mpt-7b-instruct/blob/main/attention.py#L245, the `past_key_values`), though it's multi-query attention (which has some discrepancy to multi-head attentino0.

**Comment by vvchernov** [comment]
@tqchen @yzh119  yes it has `past_key_values` as general solution, but it is managed by hyperparameter. But for mpt-7b-instruct (which is license-friendly) the `use_cache=fasle`. I propose to use my fix for convenient tool for kv cache free models. After test such model they can be improved to the model using kv cache

**Comment by yzh119** [comment]
> @tqchen @yzh119 yes it has `past_key_values` as general solution, but it is managed by hyperparameter. But for mpt-7b-instruct (which is license-friendly) the `use_cache=fasle`. I propose to use my fix for convenient tool for kv cache free models. After test such model they can be improved to the model using kv cache

I noticed that, but I suppose `use_cache=false` should only be used in training. Every transformer-based architecture should benefit from kv-cache during inference.

For SSM or RWKV, the kv-cache can still be used (but not in the way it used to be) as "memory".

**Comment by vvchernov** [comment]
Hello @yzh119! As you mentioned rwkv uses `create_kv_cache` but not for kv cache there is note that this function is used as quick hack. The MPT does not need this one for any hacks. The world is not ideal and not all transformer-based models use kv cache for inference. I need to debug the original MPT when it will be ready we can speak about its improvement with kv cache and other things. This fix is needed to check models without kv cache and it does not affect current pipeline for models with kv cache.

@tqchen are there any real reasons to not pull this fix in the code?

**Comment by vvchernov** [comment]
Hello @tqchen! Thank you for detailed answer! Yes, I've observed that there are more problems that I've seen before. Nevertheless my plan is to continue developed MPT implementation on Relax (it has more differences with LLaMa than I assumed) and mlc_chat_cli pipeline to do it more flexible for MPT specifics. I hope on your help in discussion how to do it more manageable and flexible and we will find more understanding with each other further as we work in the same boat.

P.S. This PR can be merged or I can continue development here to fix all issues related to kv cache absence as you wish

**Comment by tqchen** [comment]
Thanks @vvchernov . 

My recommendation would be focusing on some of the more unit-validation like https://github.com/mlc-ai/mlc-llm/blob/main/tests/debug/dump_intermediate.py . After that is validated, we can add it to `mlc_chat_cli` with kv cache support.

`mlc_chat_cli` is not sufficient for doing debugging and likely debugging directly the low level APIs in python would be better, via more unit-validation like https://github.com/mlc-ai/mlc-llm/blob/main/tests/debug/dump_intermediate.py

Given the current logic still can cause problems in non-kv cache version, I would suggest we close it for now and instead focusing on ramping up the tests/debug folder python function to help with some of the milestone tracking, like prefill without kv cache run e2e with logits match, before we move on to kv. 

Thank you for working on MPT, and clear communication, happy to discuss further as well
